codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2FrameReader.java

```diff
@@ -128,44 +127,40 @@ public void close() {
     @Override
     public void readFrame(ChannelHandlerContext ctx, ByteBuf input, Http2FrameListener listener)
             throws Http2Exception {
+        if (readError) {
+            input.skipBytes(input.readableBytes());
+            return;
+        }
         try {
-            while (input.isReadable()) {
-                switch (state) {
-                    case FRAME_HEADER:
-                        processHeaderState(input);
-                        if (state == State.FRAME_HEADER) {
-                            // Wait until the entire header has arrived.
-                            return;
-                        }
-
-                        // The header is complete, fall into the next case to process the payload.
-                        // This is to ensure the proper handling of zero-length payloads. In this
-                        // case, we don't want to loop around because there may be no more data
-                        // available, causing us to exit the loop. Instead, we just want to perform
-                        // the first pass at payload processing now.
-                    case FRAME_PAYLOAD:
-                        processPayloadState(ctx, input, listener);
-                        if (state == State.FRAME_PAYLOAD) {
-                            // Wait until the entire payload has arrived.
-                            return;
-                        }
-                        break;
-                    case ERROR:
-                        input.skipBytes(input.readableBytes());
+            do {
+                if (readingHeaders) {
+                    processHeaderState(input);
+                    if (readingHeaders) {
+                        // Wait until the entire header has arrived.
                         return;
-                    default:
-                        throw new IllegalStateException("Should never get here");
+                    }
                 }
-            }
+
+                // The header is complete, fall into the next case to process the payload.
+                // This is to ensure the proper handling of zero-length payloads. In this
+                // case, we don't want to loop around because there may be no more data
+                // available, causing us to exit the loop. Instead, we just want to perform
+                // the first pass at payload processing now.
+                processPayloadState(ctx, input, listener);
+                if (!readingHeaders) {
+                    // Wait until the entire payload has arrived.
+                    return;
+                }
+            } while (input.isReadable());
         } catch (Http2Exception e) {
-            state = State.ERROR;
+            readError = !Http2Exception.isStreamError(e);
```

Comment:
As I suspected there is some memory savings (on linux):

current approach with 2 booleans = 40 bytes
previous approach with enum = 48 bytes